rmw_fastrtps_shared_cpp/src/publisher.cpp

```diff
@@ -35,33 +35,17 @@ destroy_publisher(
   CustomParticipantInfo * participant_info,
   rmw_publisher_t * publisher)
 {
-  if (!publisher) {
-    RMW_SET_ERROR_MSG("publisher handle is null");
-    return RMW_RET_ERROR;
-  }
-  if (publisher->implementation_identifier != identifier) {
-    RMW_SET_ERROR_MSG("publisher handle not from this implementation");
-    return RMW_RET_ERROR;
-  }
-  if (!participant_info) {
-    RMW_SET_ERROR_MSG("participant_info is null");
-    return RMW_RET_ERROR;
-  }
+  assert(publisher->implementation_identifier == identifier);
+  static_cast<void>(identifier);
```

Comment:
In my view, there are 2 cases here:

* `rmw_fastrtps_shared_cpp` is considered a private implementation detail of `rmw_fastrtps_cpp` and `rmw_fastrtps_dynamic_cpp`.  If that is the case, then we don't need to check the identifier here at all since we are checking at the "public" API layer before calling this.  We could also not pass the identifier into this method.
* `rmw_fastrtps_shared_cpp` is considered a public library.  In that case, we should keep the checks in place and return the error message (leaving the code that was here, or maybe using the macro instead).

The same goes for the asserts below.

Reply:
So, I'll being saying that we don't consistently use either approach. It's currently a bit of both.

Considering that `rmw_fastrtps_shared_cpp` is not a standalone `rmw` implementation but the parts of one, I'm inclined towards using preconditions as much as it is possible i.e. to assert instead of checking. But sometimes, some checks can be pushed down to `rmw_fastrtps_shared_cpp` to avoid duplication. All this essentially couples all three packages' implementations for the sake of run-time efficiency and code reuse.

If we don't like that mix and match, that's fine. We can settle for either: all preconditions OR all checks in `rmw_fastrtps_shared_cpp`. Personally, I'd stick to all preconditions.

Reply:
> Considering that rmw_fastrtps_shared_cpp is not a standalone rmw implementation but the pieces of one, I'm inclined towards using preconditions as much as it is possible i.e. to assert instead of checking. But sometimes, some checks can be pushed down to rmw_fastrtps_shared_cpp to avoid code duplication.

:+1: